lib/telemetry_metrics.ex

```diff
@@ -1,5 +1,377 @@
+  information.
+
+  ## Example
+
+      last_value(
+        "vm.memory.total",
+        description: "Total amount of memory allocated by the Erlang VM", unit: :byte
+      )
+  """
+  @spec last_value(event_name(), last_value_options()) :: LastValue.t()
+  def last_value(event_name, options) do
+    {metric_name, options} = Keyword.pop(options, :name, event_name)
+    event_name = validate_event_or_metric_name!(event_name)
+    metric_name = validate_event_or_metric_name!(metric_name)
+    validate_metric_options!(options)
+    options = Keyword.merge(default_metric_options(), options)
+
+    %LastValue{
+      name: metric_name,
+      event_name: event_name,
+      metadata: options |> Keyword.fetch!(:metadata) |> metadata_spec_to_function(),
+      tags: Keyword.fetch!(options, :tags),
+      description: Keyword.fetch!(options, :description),
+      unit: Keyword.fetch!(options, :unit)
+    }
+  end
+
+  # Helpers
+
+  @spec validate_event_or_metric_name!(term()) ::
+          normalized_metric_name() | Telemetry.event_name() | no_return()
+  defp validate_event_or_metric_name!(list) when is_list(list) do
+    if Enum.all?(list, &is_atom/1) do
+      list
+    else
+      raise ArgumentError,
+            "Expected event or metric name to be a list of atoms or a string, " <>
+              "got #{inspect(list)}"
+    end
+  end
+
+  defp validate_event_or_metric_name!(event_or_metric_name)
+       when is_binary(event_or_metric_name) do
+    segments = String.split(event_or_metric_name, ".")
+
+    if Enum.any?(segments, &(&1 == "")) do
+      Logger.warn(fn ->
+        "Event or metric name #{event_or_metric_name} contains leading, trailing or " <>
+          "consecutive dots"
+      end)
+    end
+
+    Enum.map(segments, &String.to_atom/1)
+  end
+
+  defp validate_event_or_metric_name!(term) do
+    raise ArgumentError,
+          "Expected event or metric name to be a list of atoms or a string, " <>
+            "got #{inspect(term)}"
+  end
+
+  @spec default_metric_options() :: [metric_option()]
+  defp default_metric_options() do
+    [
+      metadata: [],
+      tags: [],
+      description: nil,
+      unit: :unit
+    ]
+  end
+
+  @spec validate_metric_options!([metric_option()]) :: :ok | no_return()
+  defp validate_metric_options!(options) do
+    if metadata = Keyword.get(options, :metadata), do: validate_metadata!(metadata)
+    if tags = Keyword.get(options, :tags), do: validate_tags!(tags)
+    if description = Keyword.get(options, :description), do: validate_description!(description)
+    if unit = Keyword.get(options, :unit), do: validate_unit!(unit)
+  end
+
+  @spec validate_metadata!(term()) :: :ok | no_return()
+  defp validate_metadata!(fun) when is_function(fun, 1) do
+    :ok
+  end
+
+  defp validate_metadata!(fun) when is_function(fun) do
+    {:arity, arity} = :erlang.fun_info(fun, :arity)
+
+    raise ArgumentError,
+          "Expected metadata fun to be a one-argument function, but the arity is #{arity}"
```

Comment:
Error messages start in lowercase :)

```suggestion
          "expected metadata fun to be a one-argument function, but the arity is #{arity}"
```